packages/apollo-engine-reporting/src/treeBuilder.ts

```diff
@@ -0,0 +1,225 @@
+import {
+  GraphQLResolveInfo,
+  GraphQLError,
+  ResponsePath,
+  responsePathAsArray,
+} from 'graphql';
+import { Trace } from 'apollo-engine-reporting-protobuf';
+
+export class EngineReportingTreeBuilder {
+  private startHrTime?: [number, number];
+  private stopped = false;
+  private rootNode = new Trace.Node();
+  private nodes = new Map<string, Trace.Node>([
+    [rootResponsePath, this.rootNode],
+  ]);
+  private rewriteError?: (err: GraphQLError) => GraphQLError | null;
+
+  public constructor(options: {
+    rewriteError?: (err: GraphQLError) => GraphQLError | null;
+  }) {
+    this.rewriteError = options.rewriteError;
+  }
+
+  public startTiming() {
+    if (this.startHrTime) {
+      throw Error('startTiming called twice!');
+    }
+    if (this.stopped) {
+      throw Error('startTiming called after stopTiming!');
+    }
+    this.startHrTime = process.hrtime();
+  }
+
+  public stopTiming(): { durationNs: number; rootNode: Trace.Node } {
+    if (!this.startHrTime) {
+      throw Error('stopTiming called before startTiming!');
+    }
+    if (this.stopped) {
+      throw Error('stopTiming called twice!');
+    }
+
+    this.stopped = true;
+    return {
+      durationNs: durationHrTimeToNanos(process.hrtime(this.startHrTime)),
```

Comment:
We've recently started taking great care to not use Node.js specific APIs, and `process.hrtime` certainly falls into that category, which (I suspect among other things, like its dependence on `protobuf`) makes `apollo-engine-reporting` one of those packages which has to be conditionally `require`'d depending on whether the environment supports it.

I don't think we should do anything right now, but just pointing out that it might be worth looking into using [`performance-now`](https://www.npmjs.com/package/performance-now) package which [falls back](https://github.com/myrne/performance-now/blob/master/src/performance-now.coffee) (CoffeeScript!)  to `performance.now` (browsers) and `Date.now()`.

Reply:
This is a good point and I'll add it as a comment. It's not a regression — we've always used process.hrtime for every JS implementation of timing, and I think it would be bad for people to use `Date.now()` for timing since it lacks appropriate precision (and clock skew control).

Reply:
I am underwhelmed by both `performance.now()` and `performance-now` for this purpose.

I think we definitely don't need to care about supporting engine reporting from browsers. But we may need to care about alternate non-Node server implementations. I can't recall anyone ever complaining that engine didn't work on their server due to lack of process.hrtime, though?

performance.now uses a double instead of a pair of integers and so it has less precision, and even the docs basically say to only expect microsecond precision.  (And browser implementations explicitly fudge the numbers to work around Spectre issues.)

performance-now has https://github.com/myrne/performance-now/blob/614099d729aef5d7bdbb85c0d1060166d63d05c7/src/performance-now.coffee#L6-L8 which well — the entire point of `process.hrtime` returning a silly pair of integers instead of a single number is that you can't precisely represent nanoseconds in a JS integer!  See my comment on durationHrTimeToNanos: after a process uptime of 104 days, this function will start losing precision.

Reply:
In other words, I'd be way more interested in a a function with the API of process.hrtime that tries to produce integer nanoseconds from performance.now, than the other way around.